controllers/workflow_controller.go

```diff
@@ -0,0 +1,454 @@
+
+	needAddComplete := false
+	if _, ok := new[addonapiv1.AddonCompleteLabel]; ok {
+		needAddComplete = true
+	}
+
+	if needAddComplete {
+		if old != nil {
+			if _, ok := old[addonapiv1.AddonCompleteLabel]; !ok {
+				//c.logger.Infof("mergeLabels add complete label.")
+				old[addonapiv1.AddonCompleteLabel] = addonapiv1.AddonCompleteTrueKey
+			}
+
+			for k, v := range old {
+				merged[k] = v
+			}
+			return
+		}
+		merged[addonapiv1.AddonCompleteLabel] = addonapiv1.AddonCompleteTrueKey
+		return
+	}
+
+	if _, ok := old[addonapiv1.AddonCompleteLabel]; ok {
+		//c.logger.Infof("mergeLabels remove complete label.")
+		delete(old, addonapiv1.AddonCompleteLabel)
+	}
+	for k, v := range old {
+		merged[k] = v
+	}
+}
+
+// add or remove addon finalizer according to new instance
+func (c *wfreconcile) mergeFinalizer(old, new []string) []string {
+	addFinalize := false
+	for _, f := range new {
+		if f == addonapiv1.FinalizerName {
+			// should add finalizer
+			addFinalize = true
+			break
+		}
+	}
+
+	if addFinalize {
+		needappend := true
+		for _, f := range old {
+			if f == addonapiv1.FinalizerName {
+				needappend = false
+			}
+		}
+		if needappend {
+			old = append(old, addonapiv1.FinalizerName)
+		}
+		//c.logger.Infof("mergeFinalizer after append addon finalizer %#v", old)
+		return old
+	}
+
+	// otherwise, remove finalizer
+	ret := []string{}
+	for _, f := range old {
+		if f == addonapiv1.FinalizerName {
+			continue
+		}
+		ret = append(ret, f)
+	}
+	//c.logger.Infof("mergeFinalizer after remove addon finalizer %#v", ret)
+	return ret
+}
+
+func (c *wfreconcile) mergeResources(res1, res2 []addonv1.ObjectStatus) []addonv1.ObjectStatus {
+	merged := []addonv1.ObjectStatus{}
+	check := make(map[string]addonv1.ObjectStatus)
+	mix := append(res1, res2...)
+	for _, obj := range mix {
+		id := fmt.Sprintf("%s-%s-%s", strings.TrimSpace(obj.Name), strings.TrimSpace(obj.Kind), strings.TrimSpace(obj.Group))
+		check[id] = obj
+	}
+	for _, obj := range check {
+		merged = append(merged, obj)
+	}
+	return merged
+}
+
+func (c *wfreconcile) removeFinalizer(addon *addonv1.Addon) {
+	if common.ContainsString(addon.ObjectMeta.Finalizers, addonapiv1.FinalizerName) {
+		addon.ObjectMeta.Finalizers = common.RemoveString(addon.ObjectMeta.Finalizers, addonapiv1.FinalizerName)
+	}
+}
+
+func (c *wfreconcile) removeFromCache(addonName string) {
+	// Remove version from cache
+	if ok, v := c.versionCache.HasVersionName(addonName); ok {
+		c.versionCache.RemoveVersion(v.PkgName, v.PkgVersion)
+	}
+}
+
+// update addon meta ojbect first, then update status
+func (c *wfreconcile) updateAddon(ctx context.Context, updated *addonv1.Addon) (*addonv1.Addon, error) {
+	var errs []error
+
+	latest := &addonv1.Addon{}
+	err := c.client.Get(ctx, types.NamespacedName{Namespace: updated.Namespace, Name: updated.Name}, latest)
+	if err != nil || latest == nil {
+		return nil, err
+	} else {
+
+		if reflect.DeepEqual(updated, latest) {
```

Comment:
Do we really need to do a `reflect.DeepEqual()` check here, the very first thing we do in Addon reconcile is apply the `.status.checksum`, if anything is changed.

Reply:
CheckSum only checks Addon.Spec, other changes objectMeta, finalizers, labels are also need be detected for changes, thats the points of DeepEqual